### PR TITLE
Adds logging and admin messaging to crafting

### PR DESCRIPTION
**Pull request by CydiaLamiales**
## About The Pull Request

Adds logs to crafting and makes a message to admins if a dangerous item is crafted.

## Why It's Good For The Game
Easier log reading and ticket investigation good.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Admin messages in chat, ignore the 'Memories' part
image

Individual logs : 

image

Logs through the F6 panel :

image

</details>

## Changelog
:cl:
code: added log_crafting
admin: Added logging to crafting items
admin: Added messages for crafting dangerous items
/:cl:

### CODE REVIEW

**Comment by CydiaLamiales** [comment]
Gonna be honest I don't know why the F6 menu shows the finished/started logs in reverse.

**Comment by theqwat** [comment]
just a thought as an old player does this monitor just the crafting menu crafting or does it do all out of hand crafting... ( like making a flame thrower though unscrewing a welder getting a rod then an igniter and finishing it)

**Comment by CydiaLamiales** [comment]
Only through the crafting menu. But, hmm, yeah that could be added.

**Comment by itsmeow** [comment]
> Gonna be honest I don't know why the F6 menu shows the finished/started logs in reverse.

Specifically designed this way so you can see more recent logs first. For actively monitoring someone this works well. The actual player logs menu is meant for in-depth reviews.

File: code/game/atoms.dm [hunk 1]
```diff
@@ -1415,6 +1415,43 @@
...
+  *
+  * 1 argument is for the user making the item
+  * 2 argument is for the item being created
+  * 3 argument is whether or not the user has finished creating it
+ */
+/proc/log_crafting(atom/user, atom/object, finished=FALSE)
+	var/masterpiece = object
+	var/message = "has [finished ? "finished":"started"] crafting [masterpiece]"
+	user.log_message(message, LOG_GAME)
+	var/dangerous_crafts = typecacheof(list( //List of items that make an announcement to the admins when crafted
```
Comment (itsmeow) [inline code/game/atoms.dm]: make this a global or static so it's not initialized every time you call the proc.

File: code/game/atoms.dm [hunk 2]
```diff
@@ -1415,6 +1415,43 @@
...
+				/obj/item/mop/sharp
+				))
+	if(is_type_in_typecache(masterpiece, dangerous_crafts) && finished == FALSE)
+		var/devious = null
+		var/mob/blacksmith = user
+		devious = locate(/datum/antagonist) in blacksmith.mind.antag_datums
+		if(isnull(devious))
+			message_admins("[ADMIN_LOOKUPFLW(user)] has crafted [masterpiece] as a non-antagonist.")
+		else
+			message_admins("[ADMIN_LOOKUPFLW(user)] has crafted [masterpiece] as a [devious].")
```
Comment (itsmeow) [inline code/game/atoms.dm]: antags should not message admins imo

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/atoms.dm (hunk 2 above)
Comment (CydiaLamiales) [reply code/game/atoms.dm]: Yeah removed that, was too much clutter.

File: code/game/atoms.dm (hunk 1 above)
Comment (CydiaLamiales) [reply code/game/atoms.dm]: Done

**Comment by CydiaLamiales** [comment]
Also please remove the 'Mapping' label, I accidentally commited my SM-to-Tesla changes here but removed them

File: code/modules/paperwork/paper.dm [hunk 3]
```diff
@@ -400,6 +400,24 @@
 	SHOULD_CALL_PARENT(FALSE)
 	return
 
+/obj/item/paper/crumpled/engineeringnotice
```
Comment (Penwin0) [inline code/modules/paperwork/paper.dm]: Im pretty sure this was meant for the Tesla PR, remove please

**Review by Penwin0 — CHANGES_REQUESTED** [review]

File: code/modules/paperwork/paper.dm (hunk 3 above)
Comment (CydiaLamiales) [reply code/modules/paperwork/paper.dm]: Guh, how did that even get there

File: code/_globalvars/lists/flavor_misc.dm [hunk 4]
```diff
@@ -226,3 +226,22 @@ GLOBAL_LIST_INIT(station_numerals, greek_letters + phonetic_alphabet + numbers_a
...
+				/obj/item/reagent_containers/food/drinks/bottle/molotov,
+				/obj/item/melee/baton/cattleprod,
+				/obj/item/pneumatic_cannon/ghetto,
+				/obj/item/flamethrower,
+				/obj/item/gun/ballistic/bow/pipe,
+				/obj/item/gun/ballistic/shotgun/doublebarrel/improvised,
+				/obj/item/chainsaw,
+				/obj/item/switchblade,
+				/obj/item/bombcore/chemical,
+				/obj/item/restraints/legcuffs/bola/,
```
Comment (itsmeow) [inline code/_globalvars/lists/flavor_misc.dm]: trailing /

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/_globalvars/lists/flavor_misc.dm (hunk 4 above)
Comment (CydiaLamiales) [reply code/_globalvars/lists/flavor_misc.dm]: Removed

File: code/_globalvars/lists/flavor_misc.dm [hunk 5]
```diff
@@ -226,3 +226,22 @@ GLOBAL_LIST_INIT(station_numerals, greek_letters + phonetic_alphabet + numbers_a
...
+				/obj/item/gun/ballistic/shotgun/doublebarrel/improvised,
+				/obj/item/chainsaw,
+				/obj/item/switchblade,
+				/obj/item/bombcore/chemical,
+				/obj/item/restraints/legcuffs/bola,
+				/obj/item/fireaxe/boneaxe,
+				/obj/item/kitchen/knife/shank,
+				/obj/structure/guillotine,
+				/obj/item/mop/sharp
+				)))
```
Comment (PowerfulBacon) [inline code/_globalvars/lists/flavor_misc.dm]: How about have a dangerous variable on the crafting recipe datum instead?

**Review by PowerfulBacon — DISMISSED** [review]

File: code/game/atoms.dm [hunk 6]
```diff
@@ -1415,6 +1415,23 @@
...
 
+/**
+  * Log for crafting items
+  *
+  * 1 argument is for the user making the item
+  * 2 argument is for the item being created
+  * 3 argument is for if admins should be notified if a non-antag crafts this
+ */
+/proc/log_crafting(atom/user, atom/object, dangerous = FALSE)
+	var/masterpiece = object
```
Comment (itsmeow) [inline code/game/atoms.dm]: why does this var even exist, it's a straight copy from `object` and isn't changed

File: code/game/atoms.dm [hunk 7]
```diff
@@ -1415,6 +1415,23 @@
...
+	var/message = "has crafted [masterpiece]"
+	user.log_message(message, LOG_GAME)
+	if(dangerous)
+		var/devious = null
```
Comment (itsmeow) [inline code/game/atoms.dm]: the devious var does not need to exist at all, you can just inline the result into the isnull expression

File: code/game/atoms.dm [hunk 8]
```diff
@@ -1415,6 +1415,23 @@
...
+  *
+  * 1 argument is for the user making the item
+  * 2 argument is for the item being created
+  * 3 argument is for if admins should be notified if a non-antag crafts this
+ */
+/proc/log_crafting(atom/user, atom/object, dangerous = FALSE)
+	var/masterpiece = object
+	var/message = "has crafted [masterpiece]"
+	user.log_message(message, LOG_GAME)
+	if(dangerous)
```
Comment (itsmeow) [inline code/game/atoms.dm]: do an early return `if(!dangerous) return`

**Review by itsmeow — DISMISSED** [review]

File: code/datums/components/crafting/recipes.dm [hunk 9]
```diff
@@ -12,6 +12,7 @@
 	var/subcategory = CAT_NONE
 	var/always_available = TRUE //Set to FALSE if it needs to be learned first.
 	var/one_per_turf = FALSE ///Should only one object exist on the same turf?
+	var/dangerous_craft = FALSE //Should admins be notified about this getting created by a non-antagonist?
```
Comment (ike709) [inline code/datums/components/crafting/recipes.dm]:
```suggestion
	var/dangerous_craft = FALSE /// Should admins be notified about this getting created by a non-antagonist?
```

File: code/game/atoms.dm [hunk 10]
```diff
@@ -1415,6 +1415,21 @@
...
+  * 3 argument is for if admins should be notified if a non-antag crafts this
+ */
+/proc/log_crafting(atom/user, atom/object, dangerous = FALSE)
+	var/message = "has crafted [object]"
+	user.log_message(message, LOG_GAME)
+	if(!dangerous)
+		return
+	var/mob/blacksmith = user
+	if(isnull(locate(/datum/antagonist) in blacksmith.mind.antag_datums))
+		message_admins("[ADMIN_LOOKUPFLW(user)] has crafted [object] as a non-antagonist.")
```
Comment (ike709) [inline code/game/atoms.dm]:
```suggestion
/proc/log_crafting(mob/blacksmith, atom/object, dangerous = FALSE)
	var/message = "has crafted [object]"
	blacksmith.log_message(message, LOG_GAME)
	if(!dangerous)
		return
	if(isnull(locate(/datum/antagonist) in blacksmith.mind?.antag_datums))
		message_admins("[ADMIN_LOOKUPFLW(user)] has crafted [object] as a non-antagonist.")
```

**Review by ike709 — APPROVED** [review]
Didn't check every `dangerous` item but the logging framework is fine

**Review by Penwin0 — APPROVED** [review]
Looked through all the dangerous items, looks reasonable

**Comment by Penwin0** [comment]
Dismissing Kat's and Bacon's reviews as they are handled

**Pull request merged**